### PR TITLE
Support for reorged logs in eth_Logs Subscription

### PR DESCRIPTION
**Pull request by Arindam2407**
Addressing issue #5677 

Updated EthFilter to create new task update_reorged_logs that is spawned when the EthFilter is initialized. The new task reacts to reorged blocks taking a clone of EthPubSub, checks impacted log filters and stores the reorged logs in the log filter in a new field called reorged_logs.

### CODE REVIEW

File: crates/rpc/rpc-types/Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,10 @@ alloy-primitives = { workspace = true, features = ["rand", "rlp", "serde"] }
 ethereum_ssz_derive = { version = "0.5", optional = true }
 ethereum_ssz = { version = "0.5", optional = true }
 
+#async
+tokio = { workspace = true, features = ["sync"] }
+tokio-util = "0.7"
+
```
Comment (mattsse) [inline crates/rpc/rpc-types/Cargo.toml:21]: we don't need this here

File: crates/rpc/rpc-types/src/eth/filter.rs [hunk 2]
```diff
@@ -245,18 +247,29 @@ impl FilterBlockOption {
...
     // https://eips.ethereum.org/EIPS/eip-234
     pub block_option: FilterBlockOption,
     /// Address
     pub address: FilterSet<Address>,
-    /// Topics (maxmimum of 4)
+    /// Topics (maximum of 4)
     pub topics: [Topic; 4],
+    /// Reorged logs
+    pub reorged_logs: Arc<Mutex<Option<Vec<Log>>>>,
 }
```
Comment (mattsse) [inline crates/rpc/rpc-types/src/eth/filter.rs:262]: this is the rpc type, this type should remain unchanged, there's no regorged logs field in the rpc object

this should be moved to the filter implementation

File: crates/rpc/rpc/src/eth/filter.rs [hunk 3]
```diff
@@ -37,15 +41,19 @@ use tracing::trace;
 const MAX_HEADERS_RANGE: u64 = 1_000; // with ~530bytes per header this is ~500kb
 
 /// `Eth` filter RPC implementation.
-pub struct EthFilter<Provider, Pool> {
+pub struct EthFilter<Provider, Pool, Events, Network> {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:44]: why do we need these additional types here?

File: crates/rpc/rpc/src/eth/filter.rs [hunk 4]
```diff
@@ -120,9 +141,74 @@ where
             is_valid
         })
     }
+
+    async fn update_reorged_logs(&self, pubsub: EthPubSub<Provider, Pool, Events, Network>) {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:145]: what does this do?
please include docs

File: crates/rpc/rpc/src/eth/filter.rs [hunk 5]
```diff
@@ -120,9 +141,74 @@ where
...
+            if let FilterKind::Log(ref mut filter) = active_filter.kind {
+                filter.reorged_logs = Arc::new(Mutex::new(Some(logs)));
+            }
+        }
+    }
+
+    /// Reacts to reorged blocks, checks impacted log filters, stores reorged logs in the log filter
+    pub async fn reorged_logs_stream(
+        &self,
+        pubsub: EthPubSub<Provider, Pool, Events, Network>,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:160]: I'm not exactly sure what this does.

this looks a bit convoluted.

how the on reorg logic should work is:
1. listen for reorged blocks:
2. acquire a lock on the shared state
3. try to find matching filters, if so put the logs into the ActiveFilter

File: crates/rpc/rpc/src/eth/filter.rs [hunk 6]
```diff
@@ -537,6 +635,8 @@ pub struct ActiveFilters {
 struct ActiveFilter {
     /// At which block the filter was polled last.
     block: u64,
+    /// Hash of the block at which the filter was polled last.
+    block_hash: Option<B256>,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:639]: this is where we'd put the shared state,

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks, 
I have a few suggestions to simplify this a bit:
* move the shared state into the ActiveFilter and not the rpc type
* no additional generics on the EthFilter type, not required because we want to spawn a new task that instead takes a clone of EthFilter

File: crates/rpc/rpc/src/eth/filter.rs (hunk 5 above)
Comment (Arindam2407) [reply crates/rpc/rpc/src/eth/filter.rs]: The temp_reorged_blocks stores consecutive reorged blocks as they come. The matching logs for the given filter are stored in reverted_logs. If the hash of one of these blocks matches with the last polled hash of the filter, we know that the filter previously served all the blocks stored in temp_reorged_blocks. We can push the reverted logs in the reorged_logs vector along with the filter id. If the last polled hash does not match before the block subscription receives a newly included block, we can simply leave it like that and clear the temp_reorged_blocks. The  filter did not serve reorged blocks.

**Review by mattsse — CHANGES_REQUESTED** [review]
sorry for the delay,
this is great progress!

could you please rebase this so we can proceed here?

**Comment by Arindam2407** [comment]
Please, could you review @mattsse ? Have made the requested changes

File: crates/rpc/rpc/src/eth/filter.rs [hunk 7]
```diff
@@ -55,13 +58,18 @@ where
...
     /// This also spawns a task that periodically clears stale filters.
-    pub fn new(
+    pub fn new<Events, Network>(
         provider: Provider,
         pool: Pool,
         eth_cache: EthStateCache,
         config: EthFilterConfig,
         task_spawner: Box<dyn TaskSpawner>,
-    ) -> Self {
+        pubsub: EthPubSub<Provider, Pool, Events, Network>,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:67]: we don't need this type, we only need events

File: crates/rpc/rpc/src/eth/filter.rs [hunk 8]
```diff
@@ -120,6 +136,83 @@ where
...
+                canon_state.expect("new block subscription never ends; qed").block_receipts()
+            })
+            .flat_map(futures::stream::iter)
+            .flat_map(move |(block_receipts, removed)| {
+                let mut reorged_logs: Vec<(FilterId, Vec<Log>)> = Vec::new();
+
+                if removed {
+                    temp_reorged_blocks.push(block_receipts);
+
+                    let filters = executor::block_on(self.active_filters().inner.lock());
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:182]: no executor block,

this needs to be refactored so we can have async code

I suggest not do the flat_map stuff and instead move all of this to async code in the while loop

File: crates/rpc/rpc/src/eth/filter.rs [hunk 9]
```diff
@@ -120,6 +136,83 @@ where
...
+
+                    filters.iter().for_each(|(id, active_filter)| {
+                        if let FilterKind::Log(ref filter) = active_filter.kind {
+                            let mut reverted_logs: Vec<Log> = Vec::new();
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:186]: we need to check here first if the filter is even affected by this, by checking the block number and only check for removed logs if the filter is at the same height

File: crates/rpc/rpc/src/eth/filter.rs [hunk 10]
```diff
@@ -120,6 +136,83 @@ where
...
+    {
+        let mut stream = self.reorged_logs_stream(pubsub).await;
+        while let Some((id, mut logs)) = stream.next().await {
+            let mut filters = executor::block_on(self.active_filters().inner.lock());
+            let active_filter =
+                filters.get_mut(&id).ok_or(FilterError::FilterNotFound(id)).unwrap();
+            let mut guard = active_filter.reorged_logs.lock().await;
+            if let Some(reorged_logs_vec) = guard.as_mut() {
+                reorged_logs_vec.append(&mut logs);
+            }
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:156]: here we should handle the Reorg events, basically the flat map logic from below

File: crates/rpc/rpc/src/eth/filter.rs [hunk 11]
```diff
@@ -537,10 +635,14 @@ pub struct ActiveFilters {
...
     /// At which block the filter was polled last.
     block: u64,
+    /// Hash of the block at which the filter was polled last.
+    block_hash: Option<B256>,
     /// Last time this filter was polled.
     last_poll_timestamp: Instant,
     /// What kind of filter it is.
     kind: FilterKind,
+    /// Reorged logs
+    reorged_logs: Arc<Mutex<Option<Vec<Log>>>>,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/filter.rs:645]: since we already lock the active filters set, we don't need a mutex here

**Review by mattsse — CHANGES_REQUESTED** [review]
terribly sorry for the late review...

this is going in the right direction, now we can simplify a few more things. lmk if you need additional pointers

File: crates/rpc/rpc/src/eth/filter.rs (hunk 7 above)
Comment (Arindam2407) [reply crates/rpc/rpc/src/eth/filter.rs:67]: can you explain this further?

Comment (mattsse) [reply crates/rpc/rpc/src/eth/filter.rs:67]: we only need the events trait not the pubsub isntance

**Pull request closed without merging**